python/ray/tune/examples/fault_tolerant_function_example.py

```diff
@@ -0,0 +1,92 @@
+import argparse
+import cloudpickle
+import numpy as np
+import time
+
+import ray
+from ray import tune
+from ray.tune import track
+
+
+class OptimusFn(object):
+    """Simulate learning."""
+
+    def __init__(self, params, max_t=10000):
+        self.params = params
+        self.noise = np.random.normal(size=max_t) * 0.005
+
+    def __call__(self, k, add_noise=True):
+        time.sleep(0.5)
+        b0, b1, b2 = self.params
+        score = (b0 * k / 100 + 0.1 * b1 + 0.5)**(-1) + b2 * 0.01
+        if add_noise:
+            return score + abs(self.noise[k])
+        else:
+            return score
+
+
+CHECKPOINT_FREQ = 10
+TRAINING_ITERS = 1000
+
+
+def train(config):
+    if track.is_pending_restore():
+        # Handle restoration.
+        checkpoint = track.restore()
+        func = cloudpickle.loads(checkpoint["func"])
+        mock_data = checkpoint["data"]
+        cur_i = checkpoint["iter"]
+        np.random.set_state(checkpoint["seed"])
+    else:
+        # Handle initialization.
+        params = [config["param1"], config["param2"], config["param3"]]
+        func = OptimusFn(params=params)
+        mock_data = open("/dev/urandom", "rb").read(1024)
+        cur_i = 0
+
+    for i in range(cur_i, TRAINING_ITERS):
+        new_loss = func(i)
+
+        if i % CHECKPOINT_FREQ == 0:
+            checkpoint = {
+                "func": cloudpickle.dumps(func),
+                "seed": np.random.get_state(),
+                "data": mock_data,
+                "iter": i,
+            }
+        else:
+            checkpoint = None
+
+        track.log(
+            mean_loss=float(new_loss),
+            mean_accuracy=(2 - new_loss) / 2,
+            **{track.SAVE: checkpoint},
```

Comment:
It is a little weird that checkpoints are part of the logs. This might not be a question for this PR though.

Reply:
yeah there's some discussion (see slack) to use callbacks instead